File: packages/php/tests/WebhooksModuleTest.php
<?php

declare(strict_types=1);

namespace Ecollect\Tests;

use Ecollect\Config;
use Ecollect\Exceptions\WebhookValidationException;
use Ecollect\Modules\SessionModule;
use Ecollect\Modules\WebhooksModule;
use Ecollect\Utils\Crypto;
use Ecollect\Utils\HttpClient;
use PHPUnit\Framework\TestCase;

class WebhooksModuleTest extends TestCase
{
    private function makeConfig(): Config
    {
        return new Config([
            'api_key'     => 'test-key',
            'ety_code'    => 123,
            'environment' => 'test',
        ]);
    }

    /** @return SessionModule&\PHPUnit\Framework\MockObject\MockObject */
    private function mockSession(string $token = 'session-token'): SessionModule
    {
        $mock = $this->createMock(SessionModule::class);
        $mock->method('getActive')->willReturn($token);
        return $mock;
    }

    private function makeModule(HttpClient $http = null, SessionModule $session = null): WebhooksModule
    {
        return new WebhooksModule(
            $this->makeConfig(),
            $http ?? $this->createMock(HttpClient::class),
            $session ?? $this->mockSession(),
        );
    }

    // ── verifyWebhookSignature ────────────────────────────────────────────────

    public function testVerifyWebhookSignatureReturnsTrueForValidSignature(): void
    {
        $module  = $this->makeModule();
        $payload = ['TicketId' => 42, 'ReturnCode' => 'SUCCESS'];
        $secret  = 'my-webhook-secret';
        $sig     = Crypto::hmacSha256((string) json_encode($payload), $secret);

        $this->assertTrue($module->verifyWebhookSignature($payload, $sig, $secret));
    }

    public function testVerifyWebhookSignatureReturnsFalseForWrongSignature(): void
    {
        $module  = $this->makeModule();
        $payload = ['TicketId' => 42, 'ReturnCode' => 'SUCCESS'];

        $this->assertFalse($module->verifyWebhookSignature($payload, 'bad-signature', 'my-webhook-secret'));
    }

    public function testVerifyWebhookSignatureReturnsFalseForTamperedPayload(): void
    {
        $module    = $this->makeModule();
        $original  = ['TicketId' => 42, 'ReturnCode' => 'SUCCESS'];
        $secret    = 'my-webhook-secret';
        $sig       = Crypto::hmacSha256((string) json_encode($original), $secret);

        $tampered = ['TicketId' => 99, 'ReturnCode' => 'SUCCESS'];
        $this->assertFalse($module->verifyWebhookSignature($tampered, $sig, $secret));
    }

    public function testVerifyWebhookSignatureIsCaseInsensitiveForHex(): void
    {
        $module  = $this->makeModule();
        $payload = ['event' => 'payment'];
        $secret  = 'secret';
        $sig     = strtoupper(Crypto::hmacSha256((string) json_encode($payload), $secret));

        $this->assertTrue($module->verifyWebhookSignature($payload, $sig, $secret));
    }

    // ── confirmWebhook ────────────────────────────────────────────────────────

    public function testConfirmWebhookReturnsParsedFields(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn(['ReturnCode' => 'SUCCESS']);

        $module  = $this->makeModule($http);
        $payload = [
            'SessionToken'    => 'wh-session-token',
            'TicketId'        => '12345',
            'ReturnCode'      => 'SUCCESS',
            'TranState'       => 'APPROVED',
            'TrazabilityCode' => 'TC-001',
            'TransValue'      => '100.00',
            'PayCurrency'     => 'COP',
            'FICode'          => '291',
            'PaymentSystem'   => '16',
        ];

        $result = $module->confirmWebhook($payload, 'session-token');

        $this->assertSame('SUCCESS', $result['returnCode']);
        $this->assertSame('12345', $result['ticketId']);
        $this->assertSame('APPROVED', $result['tranState']);
        $this->assertSame('TC-001', $result['trazabilityCode']);
        $this->assertSame('COP', $result['payCurrency']);
        $this->assertSame('291', $result['fiCode']);
        $this->assertSame('16', $result['paymentSystem']);
    }

    public function testConfirmWebhookThrowsWhenSessionTokenMissing(): void
    {
        $this->expectException(WebhookValidationException::class);
        $this->expectExceptionMessageMatches('/SessionToken/i');

        $module = $this->makeModule();
        $module->confirmWebhook(['TicketId' => '123']);
    }

    public function testConfirmWebhookThrowsWhenTicketIdMissing(): void
    {
        $this->expectException(WebhookValidationException::class);
        $this->expectExceptionMessageMatches('/TicketId/i');

        $module = $this->makeModule();
        $module->confirmWebhook(['SessionToken' => 'wh-token']);
    }

    public function testConfirmWebhookThrowsWhenApiReturnsFail(): void
    {
        $http = $this->createMock(HttpClient::class);
        $http->method('post')->willReturn(['ReturnCode' => 'FAIL_SESSIONNOTFOUND']);

        $module = $this->makeModule($http);

        $this->expectException(\Exception::class);
        $module->confirmWebhook(['SessionToken' => 'bad-token', 'TicketId' => '1']);
    }

    // ── buildWebhookResponse ──────────────────────────────────────────────────

    public function testBuildWebhookResponseSuccess(): void
    {
        $resp = WebhooksModule::buildWebhookResponse(true);
        $this->assertSame('SUCCESS', $resp['ReturnCode']);
    }

    public function testBuildWebhookResponseFailure(): void
    {
        $resp = WebhooksModule::buildWebhookResponse(false);
        $this->assertSame('FAIL_SYSTEM', $resp['ReturnCode']);
    }
}

File: packages/kotlin/src/test/kotlin/com/ecollect/sdk/WebhooksModuleTest.kt
package com.ecollect.sdk

import com.ecollect.sdk.exceptions.AuthenticationException
import com.ecollect.sdk.exceptions.WebhookValidationException
import com.ecollect.sdk.modules.SessionModule
import com.ecollect.sdk.modules.WebhooksModule
import com.ecollect.sdk.types.GetTransactionResponse
import com.ecollect.sdk.utils.HttpClient
import io.mockk.*
import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class WebhooksModuleTest {

    private lateinit var mockWebServer: MockWebServer
    private lateinit var httpClient: HttpClient

    @Before
    fun setUp() {
        mockWebServer = MockWebServer()
        mockWebServer.start()
        httpClient = HttpClient()
    }

    @After
    fun tearDown() {
        mockWebServer.shutdown()
    }

    private fun makeConfig(sessionToken: String = "active-session-token") = Config(
        etyCode = 123,
        sessionToken = sessionToken,
        environment = Environment.TEST,
        testBaseUrl = mockWebServer.url("/").toString()
    )

    private fun makeModule(sessionToken: String = "active-session-token"): WebhooksModule {
        val config = makeConfig(sessionToken)
        val session = mockk<SessionModule>()
        coEvery { session.getSessionToken() } returns sessionToken
        every { session.invalidate() } just Runs
        return WebhooksModule(config, httpClient, session)
    }

    private fun hmacSha256Hex(message: String, secret: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret.toByteArray(), "HmacSHA256"))
        return mac.doFinal(message.toByteArray()).joinToString("") { "%02x".format(it) }
    }

    // ── verifyWebhookSignature — HMAC path ───────────────────────────────────

    @Test
    fun `verifyWebhookSignature returns true with valid HMAC`() = runTest {
        // verifyWebhookSignature calls verifyViaApi after HMAC check passes
        mockWebServer.enqueue(MockResponse().setBody("""{"ReturnCode":"SUCCESS"}""").setResponseCode(200))

        val rawBody = """{"ReturnCode":"SUCCESS"}"""
        val secret  = "webhook-secret"
        val sig     = hmacSha256Hex(rawBody, secret)
        val payload = GetTransactionResponse(returnCode = "SUCCESS", sessionToken = "wh-session")

        val module = makeModule()
        val result = module.verifyWebhookSignature(payload, webhookSecret = secret, rawBody = rawBody, signature = sig)
        assertTrue(result)
    }

    @Test
    fun `verifyWebhookSignature throws WebhookValidationException for bad HMAC`() = runTest {
        // Bad HMAC is rejected before any HTTP call — no server response needed
        val payload = GetTransactionResponse(returnCode = "SUCCESS", sessionToken = "wh-session")
        val module = makeModule()

        try {
            module.verifyWebhookSignature(
                payload, webhookSecret = "secret", rawBody = """{"ok":true}""", signature = "bad-sig"
            )
            fail("Expected WebhookValidationException")
        } catch (e: WebhookValidationException) {
            assertTrue(e.message?.contains("HMAC") == true)
        }
    }

    @Test
    fun `verifyWebhookSignature throws when SessionToken missing from payload`() = runTest {
        // SessionToken check happens before HTTP call — no server response needed
        val payload = GetTransactionResponse(returnCode = "SUCCESS", sessionToken = null)
        val module = makeModule()

        try {
            module.verifyWebhookSignature(payload)
            fail("Expected WebhookValidationException")
        } catch (e: WebhookValidationException) {
            assertTrue(e.message?.contains("SessionToken") == true)
        }
    }

    // ── confirmWebhook ────────────────────────────────────────────────────────

    @Test
    fun `confirmWebhook returns true when API returns SUCCESS`() = runTest {
        mockWebServer.enqueue(MockResponse().setBody("""{"ReturnCode":"SUCCESS"}""").setResponseCode(200))
        val module = makeModule()

        val result = module.confirmWebhook("wh-session-token", ticketIdToVerify = 12345L)
        assertTrue(result)
    }

    @Test
    fun `confirmWebhook throws WebhookValidationException for FAIL_SESSIONNOTFOUND`() = runTest {
        mockWebServer.enqueue(MockResponse().setBody("""{"ReturnCode":"FAIL_SESSIONNOTFOUND"}""").setResponseCode(200))
        val module = makeModule()

        try {
            module.confirmWebhook("bad-token")
            fail("Expected WebhookValidationException")
        } catch (e: WebhookValidationException) {
            assertTrue(e.message?.contains("does not exist") == true)
        }
    }

    @Test
    fun `confirmWebhook throws WebhookValidationException for FAIL_TICKETIDNOTMATCH`() = runTest {
        mockWebServer.enqueue(MockResponse().setBody("""{"ReturnCode":"FAIL_TICKETIDNOTMATCH"}""").setResponseCode(200))
        val module = makeModule()

        try {
            module.confirmWebhook("wh-token", ticketIdToVerify = 999L)
            fail("Expected WebhookValidationException")
        } catch (e: WebhookValidationException) {
            assertTrue(e.message?.contains("TicketId") == true)
        }
    }

    @Test
    fun `confirmWebhook throws AuthenticationException for FAIL_ACCESSDENIED`() = runTest {
        mockWebServer.enqueue(MockResponse().setBody("""{"ReturnCode":"FAIL_ACCESSDENIED"}""").setResponseCode(200))
        val module = makeModule()

        try {
            module.confirmWebhook("wh-token")
            fail("Expected AuthenticationException")
        } catch (e: AuthenticationException) {
            assertNotNull(e)
        }
    }
}
